fix(lsp): decode 'null' in server responses as vim.NIL

arkriny opened the pull request:
The current LSP parser treats absent fields and explicitly `null` fields the same, collapsing both into nil.

One possible solution would be to remove the 'object' flag and do explicit presence checks in all LSP handlers. But this would introduce a lot of unnecessary boilerplate in places where it doesn't matter, since the distinction between `null` and missing only matters for a few specific LSP types.

The fix I currently have in a draft state is to track `?|null` paths in a static tree, which is (ideally) generated from the spec. This allows handling of certain key paths explicitly and emit `vim.NIL` for `null`, keeping nil reserved for missing fields. This approach works and fixes #31368 and unblocks #34838.

However, it comes with some drawbacks: the parser doesn't know the actual expected type when encountering a key path like "result.data"; this path could appear in multiple methods, not just CodeAction responses.


What would be ideal is to make the parser type-aware. That is:

- Parse the top-level response object.

- Use the "id" to find the method name that was originally sent.

- With that context, parse the "result" or "error" field according to the expected schema for that method.

This is how it typically works in statically typed languages: the response handler already knows the expected structure, since it needs to allocate a concrete struct. In Lua, we don’t have static typing, and as far as I know, there’s no native way to enforce such schemas during deserialization.

A type-aware parser would not only solve the `null` vs `undefined` distinction cleanly, but also allow validating responses and reporting errors early if the server breaks the spec contract. It would give us fully type-safe LSP handlers.

At this point, I’m not sure in which direction I should continue. Should I explore building a proper type-aware parser, or focus on improving the current approach by generating the path tree from the spec and documenting it better?

I feel like I might be missing something, so I'd really appreciate any feedback, guidance, or alternative ideas.

skewb1k commented:
CC @justinmk @MariaSolOs

MariaSolOs commented on runtime/lua/vim/_meta/json.lua:
```diff
@@ -30,6 +30,10 @@ vim.json = {}
 ---                                   to Lua `nil` instead of |vim.NIL|.
 ---                                 - array: (boolean) When true, converts `null` in JSON arrays
 ---                                   to Lua `nil` instead of |vim.NIL|.
+---                                 - lsp: (boolean) Similar to `object`, but applies
```
Should this behavior be specific to LSP? I can see it being helpful in other scenarios.

skewb1k replied:
It could be unified, but then there should be a way to use alternative trees. I can also imagine passing the tree dynamically from Lua instead of defining it statically. However, as I mentioned earlier, I'm not sure how suitable this approach really is, since keys can collide between different types.

mfussenegger replied:
> But I'm concerned that we don't know how many servers actually rely on this client behavior. I can't imagine every server adapting quickly after users start seeing error messages on every LSP request after Neovim update

We can use the dev cycle to learn that. Worst case we revert the commit before release if it would leave too many non-compliant servers broken

skewb1k commented:
Another idea I have is to add the ability to pass a list of keys to `decode()` that should not be decoded - leaving the corresponding nodes as-is in the final object (like json.RawMessage in Go). It would also make it possible to ignore decoding for fields like "data". For types with nullable fields, this would let us handle them specifically, since the expected type is already known. It’s a middle ground between the current parser and a fully type-aware parser that traverses the full expected schema tree.

arkriny committed:
fix(lsp): decode 'null' in server responses as vim.NIL

Fixes #31368

Problem:
Previously, 'null' value in LSP responses were decoded as 'nil'.
This caused ambiguity for fields typed as '? | null' and led to
loss of explicit 'null' values, particularly in 'data' parameters.

Solution:
Decode all JSON 'null' values as 'vim.NIL' and adjust handling
where needed. This better aligns with the LSP specification,
where 'null' and absent fields are distinct, and 'null' should
not be used to represent missing values.

This also enables proper validation of response messages to
ensure that exactly one of 'result' or 'error' is present, as
required by the JSON-RPC specification.

skewb1k commented:
test failing for some reason.. cant reproduce locally. No idea why typval.c is affected

mfussenegger reviewed (commented):
Does `documentSelector` need some additional vim.NIL handling, or are all current nil checks fine?

Other than that lgtm, although I think a news.txt entry would be justified as this affects all plugins using any of the properties that are defined with `|null`

ribru17 commented on runtime/doc/news.txt:
```diff
@@ -28,6 +28,7 @@ EXPERIMENTS
 
 LSP
 
+• Absence of values in LSP messages is now represented as |vim.NIL| instead of nil.
```
I could be wrong, but I think this belongs in the regular breaking changes section, rather than breaking changes in head.

skewb1k commented:
Rebased on master and fixed result handling. The code appears stable in daily use and looks complete for this PR. Let me know if I should adjust documentation, commit messages, news.txt, or test cases.

There seems to be an issue with oldtest checks (which you are likely already aware of).

justinmk commented:
rebasing will fix the oldtest ci

skewb1k commented:
> Now that this PR has been merged, should I consider it necessary to explicitly handle `vim.NIL` in LSP-related plugins?     

Basically yes - if a field is defined as nullable in the spec, you should now handle `vim.NIL`.

> `if response.items then ... end` ↓ `if response.items and response.items ~= vim.NIL then ... end`

If items is both nullable and optional, then yes. But it's likely just one of the two. If a server confuses these properties, it's not spec-compliant and shouldn't need to be handled.

hrsh7th commented:
The example code is written with backward compatibility in mind. If the plugin author wanted to make it work with both neovim stable and neovim nightly, an if statement like this would be necessary.

I will prepare a utility within the plugin to make the determination. Thank you for your reply.
